### PR TITLE
Fix Aria scan for IN filters

### PR DESCRIPTION
**Pull request by mbasmanova**
This set of changes is fixing scan for NOT IN and IN with large number of values. TestHiveDistributedQueries#testLargeIn is now passing.

```
SELECT orderkey FROM orders WHERE orderkey NOT IN (1, 2, 3);

SELECT orderkey FROM orders WHERE orderkey IN (1, 2, 3,...5000);
```

### CODE REVIEW

**Comment by oerling** [comment]
There could be a bug in a case where 

 

C is null or (c < 10 and c > 11)

 

The range is always false but the resulting filter would be always false instead of is null.

 

Is there a TupleDomain representation for c not in (1, 2, 3) ?

**Comment by mbasmanova** [comment]
@oerling 

> Is there a TupleDomain representation for c not in (1, 2, 3) ?

This is represented as c != 1 and c != 2 and c != 3, e.g. all - 1 - 2 -3. See com.facebook.presto.sql.planner.TestDomainTranslator#testInOptimization

```
        testDomain = Domain.create(
                ValueSet.ofRanges(
                        Range.lessThan(BIGINT, 4L))
                        .intersect(ValueSet.all(BIGINT)
                                .subtract(ValueSet.ofRanges(Range.equal(BIGINT, 1L), Range.equal(BIGINT, 2L), Range.equal(BIGINT, 3L))))
                        .union(ValueSet.ofRanges(Range.range(BIGINT, 7L, true, 9L, true))), false);

        tupleDomain = withColumnDomains(ImmutableMap.<Symbol, Domain>builder().put(C_BIGINT, testDomain).build());
        assertEquals(toPredicate(tupleDomain), or(and(lessThan(C_BIGINT, bigintLiteral(4L)), not(in(C_BIGINT, ImmutableList.of(1L, 2L, 3L)))), between(C_BIGINT, bigintLiteral(7L), bigintLiteral(9L))));
```

**Comment by mbasmanova** [comment]
@oerling

> There could be a bug in a case where C is null or (c < 10 and c > 11)

I tested this and I saw that the filter was is-null.

**Comment by mbasmanova** [comment]
@oerling 

> There could be a bug in a case where C is null or (c < 10 and c > 11)

This particular case works fine, but you are correct that the original version was losing `nullAllowed`. I fixed that:

```
                if (rangeFilters.isEmpty()) {
                        filter = nullAllowed ? Filters.isNull() : Filters.alwaysFalse();
                    }
                    else {
                        filter = Filters.createMultiRange(rangeFilters, nullAllowed);
                    }
```

File: presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java [hunk 1]
```diff
@@ -277,43 +278,45 @@ else if (expression instanceof LambdaDefinitionExpression) {
...
+                        }
+                        else {
+                            v.add(conjunct);
+                        }
+                        return v;
+                    });
+        }
+
+        for (List<RowExpression> conjuncts : inputsToConjuncts.values()) {
+            RowExpression firstConjunct = conjuncts.get(0);
```
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java]: Should there be a check that conjuncts is not empty?

File: presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java [hunk 2]
```diff
@@ -277,43 +278,45 @@ else if (expression instanceof LambdaDefinitionExpression) {
...
-                }
-                result.add(pageFunctionCompiler.compileFilter(conjunct, classNameSuffix));
+
+        ImmutableList.Builder<Supplier<PageFilter>> result = ImmutableList.builder();
+        CallExpression topAnd = (CallExpression) filter.get();
+        Map<Set<InputReferenceExpression>, List<RowExpression>> inputsToConjuncts = new HashMap();
+        for (RowExpression conjunct : allConjuncts) {
+            ImmutableSet.Builder<InputReferenceExpression> inputs = ImmutableSet.builder();
+            collectInputs(conjunct, inputs);
+            inputsToConjuncts.compute(inputs.build(),
```
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java]: What do you think of this:
`inputsToConjuncts.computeIfAbsent(inputs.build(), k -> new ArrayList<>).add(conjunct);`

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java [hunk 3]
```diff
@@ -188,6 +224,39 @@ public boolean isEquality()
...
         }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: Why not use instanceof? It returns false if o == null:
`if (o instanceof BigIntRange)`

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java [hunk 4]
```diff
@@ -444,26 +513,29 @@ public void addMember(SubfieldPath.PathElement member, Filter filter)
...
-            if (this.filters[0] instanceof BigintRange) {
-                longLowerBounds = new long[this.filters.length];
-                for (int i = 0; i < this.filters.length; i++) {
-                    BigintRange range = (BigintRange) this.filters[i];
-                    longLowerBounds[i] = range.getLower();
-                    if (i > 0 && longLowerBounds[i] < ((BigintRange) this.filters[i - 1]).getUpper()) {
-                        throw new IllegalArgumentException("Bigint filter range set must be in ascending order of lower bound and ranges must be disjoint");
-                    }
+            this.filters = filters.toArray(new Filter[0]);
+            if (filters.get(0) instanceof BigintRange) {
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: Should there be a not empty check for filters before the get(0)?

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java [hunk 5]
```diff
@@ -503,25 +575,59 @@ public boolean testBytes(byte[] buffer, int offset, int length)
...
         }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: Why not use instanceof here as well?

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java [hunk 6]
```diff
@@ -503,25 +575,59 @@ public boolean testBytes(byte[] buffer, int offset, int length)
...
 
-    public static class InTest
+    public static class BingintValues
             extends Filter
     {
-        static final long emptyMarker = 0xdeadbeefbadefeedL;
-        static final long M = 0xc6a4a7935bd1e995L;
-        private long[] longs;
-        int size;
+        private static final long emptyMarker = 0xdeadbeefbadefeedL;
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: Is there a chance that a value can equal the emptyMarker?

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java [hunk 7]
```diff
@@ -557,12 +663,45 @@ public boolean testLong(long value)
...
         }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: Can you use instanceof here as well?

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainOrcPredicate.java [hunk 8]
```diff
@@ -357,8 +357,16 @@ else if (ranges.size() == 1) {
                     filter = createRangeFilter(type, ranges.get(0), nullAllowed);
                 }
                 else {
-                    List<Filter> rangeFilters = ranges.stream().map(r -> createRangeFilter(type, r, false)).collect(toList());
-                    filter = Filters.createMultiRange(rangeFilters, nullAllowed);
+                    List<Filter> rangeFilters = ranges.stream()
+                            .map(r -> createRangeFilter(type, r, false))
+                            .filter(f -> f != Filters.alwaysFalse())
+                            .collect(toList());
```
Comment (elonazoulay) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainOrcPredicate.java]: .collect(toImmutableList())

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java [hunk 9]
```diff
@@ -444,26 +513,29 @@ public void addMember(SubfieldPath.PathElement member, Filter filter)
...
+                this.longLowerBounds = filters.stream()
+                        .mapToLong(filter -> ((BigintRange) filter).getLower())
+                        .toArray();
+
+                long[] upperBounds = filters.stream()
+                        .mapToLong(filter -> ((BigintRange) filter).getUpper())
+                        .toArray();
+
+                for (int i = 1; i < longLowerBounds.length; i++) {
+                    checkArgument(longLowerBounds[i] >= upperBounds[i - 1], "bigint ranges must not overlap");
```
Comment (tdcmeehan) [inline presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: They may not be in ascending order?

File: presto-orc/src/test/java/com/facebook/presto/orc/TestTupleDomainOrcPredicateFilters.java [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+
+    @BeforeClass
+    public void setup()
+    {
+        metadata = createTestMetadataManager();
+        literalEncoder = new LiteralEncoder(metadata.getBlockEncodingSerde());
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
```
Comment (tdcmeehan) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestTupleDomainOrcPredicateFilters.java]: Why do we need to do this?

**Comment by oerling** [comment]
BigintRange could be declared final. Besides, if o were an instance of a subclass of BigintRange other than the class of this, the test would fail.

File: presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java (hunk 1 above)
Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java]: @elonazoulay The above loop never creates empty lists. Are you suggesting to add a verify call here?

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java (hunk 6 above)
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: @elonazoulay Yes, that's possible. There is extra logic to handle this case.

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java (hunk 9 above)
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: They shouldn't be.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestTupleDomainOrcPredicateFilters.java (hunk 10 above)
Comment (mbasmanova) [reply presto-orc/src/test/java/com/facebook/presto/orc/TestTupleDomainOrcPredicateFilters.java]: We don't. Removing.

**Comment by mbasmanova** [comment]
@tdcmeehan @elonazoulay @oerling Folks, thank you for reviewing. I addressed the comments and added a commit that fixes queries with IN predicates with large number of values. Please, take another look.

File: presto-orc/src/main/java/com/facebook/presto/orc/Filters.java (hunk 6 above)
Comment (tdcmeehan) [reply presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: nit: `EMPTY_MARKER`

Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/Filters.java]: @tdcmeehan Updated

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/ValueSet.java [hunk 11]
```diff
@@ -65,7 +65,7 @@ static ValueSet of(Type type, Object first, Object... rest)
         throw new IllegalArgumentException("Cannot create discrete ValueSet with non-comparable type: " + type);
     }
 
-    static ValueSet copyOf(Type type, Collection<Object> values)
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/predicate/ValueSet.java]: @tdcmeehan This didn't work, hence, I changed it. However, I see that I no longer use this method. I'll revert the change.

**Pull request merged**